### PR TITLE
pythonPackages311.pysha3: replace with safe-pysha3

### PR DESCRIPTION
**Pull request by nevivurn**
###### Description of changes

- Package is abandoned and vulnerable: tiran/pysha3#29
- It also [breaks builds](https://hydra.nixos.org/job/nixos/trunk-combined/nixpkgs.python311Packages.pysha3.x86_64-linux) due to tiran/pysha3#30
- `safe-pysha3` is a drop-in replacement
  - But the package name is different and may break builds, not sure if I should delete the alias?
  - Also, `safe-pysha3` no longer supports Python < 3.9.

ZHF: #230712

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by nevivurn** [comment]
Right, assuming the packages would have otherwise been built successfully, this breaks builds for python < 3.9. This only affects eth-hash as far as I can tell.

**Comment by nevivurn** [comment]
Sure thing. For future reference, is it written out anywhere which python versions are supported for a given release of nixpkgs / when it is OK to remove non-building packages?

**Comment by wegank** [comment]
I don't think that's documented, but you can usually figure it out by looking in `pkgs/top-level/all-packages.nix`:

```
  python27Packages = python27.pkgs;
  python38Packages = python38.pkgs;
  python39Packages = python39.pkgs;
  python310Packages = recurseIntoAttrs python310.pkgs;
  python311Packages = recurseIntoAttrs python311.pkgs;
  python312Packages = python312.pkgs;
```

Only the package sets with `recurseIntoAttrs` are built and cached.

**Comment by SuperSandro2000** [comment]
> * But the package name is different and may break builds, not sure if I should delete the alias?

Is the import name the same?

**Comment by nevivurn** [comment]
> Is the import name the same?

Yes, both `pysha3` and `safe-pysha3` provide `sha3` (and monkey-patch `hashlib`). Might require some patching in `setup.py` like https://github.com/NixOS/nixpkgs/blob/5c8385332672f54002e21c5fdd477c9ee50de876/pkgs/development/python-modules/manticore/default.nix#L61

**Comment by nevivurn** [comment]
rebased due to conflicts from #232892

File: pkgs/top-level/python-aliases.nix [hunk 1]
```diff
@@ -225,6 +225,7 @@ mapAliases ({
   Pyro4 = pyro4; # added 2023-02-19
   Pyro5 = pyro5; # added 2023-02-19
   PyRSS2Gen = pyrss2gen; # added 2023-02-19
+  pysha3 = safe-pysha3; # added 2023-05-19
```
Comment (wegank) [inline pkgs/top-level/python-aliases.nix:228]: Probably just deprecate it instead, if additional postPatch is needed.

Comment (nevivurn) [reply pkgs/top-level/python-aliases.nix]: Done

**Review by wegank — APPROVED** [review]

**Pull request merged**